pkg/services/blockchain/eth/resolver.go

```diff
@@ -148,9 +245,206 @@ func (r *Resolver) Resolve(
+				errors.New("no wallet key found for generating signature")
+		}
+		primaryType := services.IdentityStateType
+		if opts.GistRoot != nil {
+			primaryType = services.GlobalStateType
+		}
+		signature, err = r.signTypedData(primaryType, did, identityState)
+		if err != nil {
+			return services.IdentityState{}, err
+		}
+	}
+
+	identityState.Signature = signature
+
 	return identityState, err
 }
 
+func (r *Resolver) VerifyState(
+	primaryType services.PrimaryType,
+	identityState services.IdentityState,
+	did w3c.DID,
+) (bool, error) {
+	walletAddress, err := r.GetWalletAddress()
+	if err != nil {
+		return false, err
+	}
+
+	typedData, err := r.TypedData(primaryType, did, identityState, walletAddress)
+	if err != nil {
+		return false, err
+	}
+
+	authData := AuthData{TypedData: typedData, Signature: identityState.Signature, Address: walletAddress}
+	return r.verifyTypedData(authData)
+}
+
+func TimeStampFn() string {
+	timestamp := strconv.FormatInt(time.Now().UTC().Unix(), 10)
+	return timestamp
+}
+
+func (r *Resolver) TypedData(primaryType services.PrimaryType, did w3c.DID, identityState services.IdentityState, walletAddress string) (apitypes.TypedData, error) {
+	id, err := core.IDFromDID(did)
+	if err != nil {
+		return apitypes.TypedData{},
+			fmt.Errorf("invalid did format for did '%s': %v", did, err)
+	}
+	ID := id.BigInt().String()
+	idType := fmt.Sprintf("0x%02X%02X", id.Type()[0], id.Type()[1])
+
+	apiTypes := apitypes.Types{}
+	message := apitypes.TypedDataMessage{}
+	primaryTypeString := ""
+	timestamp := TimeStamp()
+
+	switch primaryType {
+	case services.IdentityStateType:
+		state := "0"
+		replacedAtTimestamp := "0"
+
+		if identityState.StateInfo != nil {
+			state = identityState.StateInfo.State.String()
+			replacedAtTimestamp = identityState.StateInfo.ReplacedAtTimestamp.String()
+		}
+		primaryTypeString = "IdentityState"
+		apiTypes = IdentityStateAPITypes
+		message = apitypes.TypedDataMessage{
+			"timestamp":           timestamp,
+			"id":                  ID,
+			"state":               state,
+			"replacedAtTimestamp": replacedAtTimestamp,
+		}
+	case services.GlobalStateType:
+		root := "0"
+		replacedAtTimestamp := "0"
+
+		if identityState.GistInfo != nil {
+			root = identityState.GistInfo.Root.String()
+			replacedAtTimestamp = identityState.GistInfo.ReplacedAtTimestamp.String()
+		}
+		primaryTypeString = "GlobalState"
+		apiTypes = GlobalStateAPITypes
+		message = apitypes.TypedDataMessage{
+			"timestamp":           timestamp,
+			"idType":              idType,
+			"root":                root,
+			"replacedAtTimestamp": replacedAtTimestamp,
+		}
+	}
+
+	typedData := apitypes.TypedData{
+		Types:       apiTypes,
+		PrimaryType: primaryTypeString,
+		Domain: apitypes.TypedDataDomain{
+			Name:              "StateInfo",
+			Version:           "1",
+			ChainId:           math.NewHexOrDecimal256(int64(0)),
+			VerifyingContract: "0x0000000000000000000000000000000000000000",
```

Comment:
let's try to avoid using hardcoded strings if it is possible, use instead ethCommon.Address{}.String()